python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py

```diff
@@ -820,6 +838,208 @@ def evaluate_for_asset(
         return AssetConditionResult.create(context, true_subset, subsets_with_metadata)
 
 
+@whitelist_for_serdes
+class SkipOnNotAllParentsUpdatedSinceCronRule(
+    AutoMaterializeRule,
+    NamedTuple(
+        "_SkipOnNotAllParentsUpdatedSinceCronRule",
+        [("cron_schedule", str), ("timezone", str)],
+    ),
+):
+    @property
+    def decision_type(self) -> AutoMaterializeDecisionType:
+        return AutoMaterializeDecisionType.SKIP
+
+    @property
+    def description(self) -> str:
+        return f"waiting until all upstream assets have updated since the last cron schedule tick of '{self.cron_schedule}' (timezone: {self.timezone})"
+
+    def passed_time_window(self, context: AssetConditionEvaluationContext) -> TimeWindow:
+        """Returns the window of time that has passed between the previous two cron ticks. All
+        parent assets must contain all data from this time window in order for this asset to be
+        materialized.
+        """
+        previous_ticks = reverse_cron_string_iterator(
+            end_timestamp=context.evaluation_time.timestamp(),
+            cron_string=self.cron_schedule,
+            execution_timezone=self.timezone,
+        )
+        end_time = next(previous_ticks)
+        start_time = next(previous_ticks)
+
+        return TimeWindow(start=start_time, end=end_time)
+
+    def get_parent_subset_updated_since_cron(
+        self,
+        context: AssetConditionEvaluationContext,
+        parent_asset_key: AssetKey,
+        passed_time_window: TimeWindow,
+    ) -> ValidAssetSubset:
+        """Returns the AssetSubset of a given parent asset that has been updated since the end of
+        the previous cron tick.
+        """
+        if (
+            # first tick of evaluating this condition
+            context.previous_evaluation_state is None
+            or context.previous_evaluation_timestamp is None
+            # new cron tick has happened since the previous tick
+            or passed_time_window.end.timestamp() > context.previous_evaluation_timestamp
+        ):
+            return context.instance_queryer.get_asset_subset_updated_after_time(
+                asset_key=parent_asset_key, after_time=passed_time_window.end
+            )
+        else:
+            # previous state still valid
+            new_parent_subset = context.instance_queryer.get_asset_subset_updated_after_cursor(
+                asset_key=parent_asset_key, after_cursor=context.previous_max_storage_id
+            )
+            previous_parent_subset = next(
+                iter(
+                    [
+                        subset
+                        for subset in context.previous_evaluation_state.get_extra_state(
+                            context.condition, list
+                        )
+                        or []
+                        if subset.asset_key == parent_asset_key
+                    ]
+                ),
+                context.empty_subset(),
+            )
```

Comment:
The chunk of code is a bit of a beast and I think worth wrapping and encapulating a bit. A couple intermediate variables would make this more trivially readable.

Wrapping and documenting that this condition gets and sets a list of subsets (e.g. is it a ValidAssetSubset or an AssetSubset?) would also be very nice. I really have to think hard when I look at this code.

Reply:
good point, updated!